packages/node_modules/@ciscospark/widget-roster/src/selector.js

```diff
@@ -1,111 +1,228 @@
   }
 );
 
-export const getRosterWidgetProps = createSelector(
-  [
-    getSparkInstance,
-    getConversation,
-    getUser,
-    getAvatars,
-    getSearchResults,
-    getWidget
-  ],
-  // eslint-disable-next-line max-params
-  (sparkInstance, conversation, user, avatars, searchResults, widgetRoster) => {
+export const getParticipantsById = createSelector(
+  [getConversation, getAvatars, getCurrentUser],
+  (conversation, avatars, currentUser) => {
     const participants = conversation.get(`participants`).toJS();
     const inFlightParticipants = conversation.get(`inFlightParticipants`).toJS();
-    const currentUser = user.get(`currentUser`);
     const isConsumerOrg = currentUser.orgId === `consumer`;
+    const participantsById = {};
 
+    participants.forEach((participant) => {
+      participant.avatar = avatars.getIn([`items`, participant.id]);
+      participant.isExternal = currentUser && !isConsumerOrg && currentUser.orgId !== `consumer` && currentUser.orgId !== participant.orgId;
+      participant.displayName = participant.displayName ? participant.displayName : participant.name;
+      participant.isPending = has(inFlightParticipants, [`removing`, participant.id]) || has(inFlightParticipants, [`adding`, participant.id]);
+      participantsById[participant.id] = participant;
+    });
+
+    return participantsById;
+  }
+);
+
+export const getParticipants = createSelector(
+  [
+    getConversation,
+    getParticipantsById,
+    getCurrentUser,
+    getAvatars,
+    getCurrentCall,
+    getFormatMessage
+  ],
+  ( // eslint-disable-line max-params
+    conversation,
+    participantsById,
+    currentUser,
+    avatars,
+    currentCall,
+    formatMessage
+  ) => {
+    const people = [];
     let hasExternalParticipants = false;
+    let count = 0;
 
-    const sortParticipants = (participantA, participantB) => {
-      const A = participantA.displayName;
-      const B = participantB.displayName;
-      // eslint-disable-next-line no-nested-ternary
-      return A < B ? -1 : A > B ? 1 : 0;
-    };
 
-    const mapParticipant = (participant) => {
-      participant.avatar = avatars.getIn([`items`, participant.id]);
-      participant.external = currentUser && !isConsumerOrg && currentUser.orgId !== participant.orgId;
-      if (participant.external) {
-        hasExternalParticipants = true;
+    if (currentCall && currentCall.getIn([`callState`, `connected`])) {
+      const callMemberships = currentCall.get(`instance`).memberships;
+      const inMeeting = [];
+      const notInMeeting = [];
+      callMemberships.forEach((m) => {
+        const participant = participantsById[m.personUuid];
+        if (participant) {
+          if (m.state === `connected`) {
+            inMeeting.push(participant);
+          }
+          else {
+            notInMeeting.push(participant);
+          }
+        }
+      });
+
+      if (inMeeting.length > 0) {
+        people.push({
+          label: formatMessage(messages.inMeeting),
```

Comment:
Still feel like `formatMessage` is gross in a selector. Is there a way we could format the message in the container before passing it on to the people list? This works for now, but maybe in our performance improvements we could figure out a way.

Reply:
Yea... it is gross. Haven't found a great way to do it yet for things like this. I'm hoping that once we do some normalizing of the reducers and stores that a better pattern will emerge.